### ISSUE TITLE
Repo: Export all rule Options and MessageId types

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#10555]
### Suggestion

Splitting out of #7670 -> #8443 -> #10554: we're going to soon have a rule that enforces all types used in a file's exports are themselves exported. That's a lot of `+ export ...` diffs. I'd send a PR to do that separately before #10554, to get that PR to be less huge.

### Additional Info

💖

**Comment by kirkwaiblinger** [comment]
Do you plan to do this as well for the rules whose type parameters are inferred, like

https://github.com/typescript-eslint/typescript-eslint/blob/0d65f179ac88ab6bb60cd3fcb1ad67bdc725a9d1/packages/eslint-plugin/src/rules/no-duplicate-enum-values.ts#L7

Asking because I'm almost curious whether we should just infer the Options/MessageId types where we need them rather than even export the types, using utilities like the following:

https://github.com/typescript-eslint/typescript-eslint/blob/2ad3404690cd601d1bbb80daa7f861ab5bb127e4/packages/utils/src/eslint-utils/InferTypesFromRule.ts#L13-L21

**Comment by JoshuaKGoldberg** [comment]
No, just the ones that would violate the rule. I think inferring types is nice but will probably take a lot longer.
